### PR TITLE
Task-Specific Submission ReRun API

### PR DESCRIPTION
**Pull request by gibsonbailey**
Feature add: Rerun a submission with a private task. This will allow clinicians to use the Codabench API to run performant submissions on data of their own without exposing their data to the public.

# @ mention of reviewers
@ckcollab 
@Logan-Ruf 

# A checklist for hand testing
- [x] Create a task which is not attached to a competition
- [x] Make a `POST` request to `/api/submissions/<int:submission_pk>/re_run_submission/?task_key=<uuid:task_key>&private=true`
- [x] Verify submission output is not visible on original competition by users other than the one who initiated the rerun.
- [x] Verify submission output is correct for new private task

# TODO:
- [x] Change `re_run_submission` action to accept a `POST` only rather than `GET`
- [x] Make sure nobody can view submission output but the requestor that reran the submission with the private dataset (write test)

# Checklist
- [x] Code review by me 
- [x] Hand tested by me 
- [x] I'm proud of my work
- [x] Code review by reviewer
- [x] Hand tested by reviewer
- [x] Ready to merge

### CODE REVIEW

File: docs/example_scripts/example_submission.py [hunk 1]
```diff
@@ -31,9 +31,9 @@
 CODALAB_URL = 'http://localhost/'
 USERNAME = 'admin'
 PASSWORD = 'admin'
-PHASE_ID = None
+PHASE_ID = 28
 TASK_LIST = []
-SUBMISSION_ZIP_PATH = '../tests/functional/test_files/submission.zip'
+SUBMISSION_ZIP_PATH = '../../src/tests/functional/test_files/submission.zip'
```
Comment (gibsonbailey) [inline docs/example_scripts/example_submission.py:36]: This path change is due to the `mv` of example_scripts into the docs directory.

File: src/apps/competitions/tasks.py [hunk 2]
```diff
@@ -255,10 +256,15 @@ def _run_submission(submission_pk, task_pks=None, is_scoring=False):
...
+    if submission.is_specific_task_re_run:
+        # Should only be one task for a specified task submission
+        tasks = Task.objects.filter(pk__in=task_pks)
+    elif task_pks is None:
+        tasks = submission.phase.tasks.all()
     else:
-        tasks = submission.phase.tasks.filter(pk__in=task_pks).order_by('pk')
+        tasks = submission.phase.tasks.filter(pk__in=task_pks)
+
+    tasks = tasks.order_by('pk')
```
Comment (gibsonbailey) [inline src/apps/competitions/tasks.py:267]: Consolidated calls to `order_by` here.

File: src/factories.py [hunk 3]
```diff
@@ -162,6 +162,7 @@ class Meta:
         created_by=factory.SelfAttribute('..owner'),
         created_when=factory.SelfAttribute('..created_when'),
     )
+    task = factory.LazyAttribute(lambda submission: random.choice(submission.phase.tasks.all()))
```
Comment (gibsonbailey) [inline src/factories.py:165]: Was having problems because my submissions didn't have tasks assigned after a `./reset_db.sh`. This fixes it!

**Comment by Logan-Ruf** [comment]
When I ran a submission against a new task. It showed up under the original competition. Not sure if this a problem
image

image

**Comment by gibsonbailey** [comment]
> When I ran a submission against a new task. It showed up under the original competition. Not sure if this a problem
I think this is ok. It's really just important that the submission does not get exposed to anyone but the requestor that initiated the re-run.

File: src/apps/api/views/submissions.py [hunk 4]
```diff
@@ -29,13 +30,17 @@ class SubmissionViewSet(ModelViewSet):
 
     def check_object_permissions(self, request, obj):
         if self.action in ['submission_leaderboard_connection']:
+            if obj.is_private or obj.is_specific_task_re_run:
+                raise PermissionDenied("Cannot add task-specific submission re-runs to leaderboards.")
```
[the remark this answers is not in the document]
Comment (gibsonbailey) [reply src/apps/api/views/submissions.py:34]: I'll take another look and consider that. I like that this is explicit, but that may be simpler which is good too.

[the remark this answers is not in the document]
Comment (gibsonbailey) [reply src/apps/api/views/submissions.py:34]: Considering also getting rid of `is_private` and just using the `is_public` field that's already on Submission. I'll have to dig into it to see what it's already being used for to evaluate if I can use it for this too.

File: src/apps/api/views/submissions.py [hunk 5]
```diff
@@ -139,16 +144,25 @@ def cancel_submission(self, request, pk):
         for child in submission.children.all():
             child.cancel()
         canceled = submission.cancel()
-
         return Response({'canceled': canceled})
 
-    @action(detail=True, methods=('GET',))
+    @action(detail=True, methods=('POST',))
```
[the remark this answers is not in the document]
Comment (gibsonbailey) [reply src/apps/api/views/submissions.py:149]: Def

File: src/apps/competitions/migrations/0024_auto_20201029_2210.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+            model_name='submission',
+            name='is_private',
+            field=models.BooleanField(default=False),
+        ),
+        migrations.AddField(
+            model_name='submission',
+            name='is_specific_task_re_run',
+            field=models.BooleanField(default=False),
+        ),
+    ]
```
Comment (Logan-Ruf) [inline src/apps/competitions/migrations/0024_auto_20201029_2210.py:23]: There's a 0024 migration already on develop. Need to bump this.

Comment (gibsonbailey) [reply src/apps/competitions/migrations/0024_auto_20201029_2210.py:23]: 👌🏼

File: src/apps/api/views/submissions.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (gibsonbailey) [reply src/apps/api/views/submissions.py:34]: Getting rid of `is_private` without replacement. It was essentially a redundant field for `is_specific_task_rerun`. If we need finer-grained control over access to `is_specific_task_rerun` submissions, we can add another field at a later date.

**Pull request merged**